### ISSUE TITLE
Cannot use Custom Expression after aggregation

### ISSUE THREAD

**Issue by flamber** [metabase/metabase#13857]
**Describe the bug**
Cannot use a Custom Expression in a Custom Column after aggregation/aggregated summarize.
Most likely a duplicate of #12839

**To Reproduce**
1. Custom question > Sample Dataset > Orders
2. Summarize: Custom Expression formula `1 * 1` as "CustomExpression" by CreatedAt:Month
3. Custom Column formula `[CustomExpression] * 1234` as "CustomColumn"
![image](https://user-images.githubusercontent.com/1447303/99802122-6acdcd00-2b37-11eb-9fbb-94e21b999530.png)
4. Errors with `Column "source.CustomExpression" not found;`

<details><summary>Full stacktrace</summary>

```
2020-11-20 13:51:59,721 ERROR middleware.catch-exceptions :: Error processing query: null
{:database_id 4,
 :started_at #t "2020-11-20T13:51:59.351889+01:00[Europe/Copenhagen]",
 :state "42S22",
 :json_query
 {:database 4,
  :query
  {:source-query {:source-table 11, :aggregation [["aggregation-options" ["*" 1 1] {:display-name "CustomExpression"}]], :breakout [["datetime-field" ["field-id" 86] "year"]]},
   :expressions {:CustomColumn ["*" ["field-literal" "CustomExpression" "type/Float"] 1234]}},
  :type "query",
  :parameters [],
  :middleware {:js-int-to-string? true, :add-default-userland-constraints? true}},
 :native
 {:query
  "SELECT \"source\".\"CREATED_AT\" AS \"CREATED_AT\", \"source\".\"expression\" AS \"expression\", \"source\".\"CustomColumn\" AS \"CustomColumn\" FROM (SELECT \"source\".\"CustomExpression\" AS \"CustomExpression\", \"source\".\"CREATED_AT\" AS \"CREATED_AT\", \"source\".\"expression\" AS \"expression\", (\"source\".\"CustomExpression\" * 1234) AS \"CustomColumn\" FROM (SELECT parsedatetime(year(\"PUBLIC\".\"ORDERS\".\"CREATED_AT\"), 'yyyy') AS \"CREATED_AT\", (1 * 1) AS \"expression\" FROM \"PUBLIC\".\"ORDERS\" GROUP BY parsedatetime(year(\"PUBLIC\".\"ORDERS\".\"CREATED_AT\"), 'yyyy') ORDER BY parsedatetime(year(\"PUBLIC\".\"ORDERS\".\"CREATED_AT\"), 'yyyy') ASC) \"source\") \"source\" LIMIT 2000",
  :params nil},
 :status :failed,
 :class org.h2.jdbc.JdbcSQLException,
 :stacktrace
 ["org.h2.message.DbException.getJdbcSQLException(DbException.java:357)"
  "org.h2.message.DbException.get(DbException.java:179)"
  "org.h2.message.DbException.get(DbException.java:155)"
  "org.h2.expression.ExpressionColumn.optimize(ExpressionColumn.java:150)"
  "org.h2.expression.Alias.optimize(Alias.java:51)"
  "org.h2.command.dml.Select.prepare(Select.java:858)"
  "org.h2.command.Parser.prepare(Parser.java:263)"
  "org.h2.engine.Session.prepare(Session.java:576)"
  "org.h2.table.TableView.compileViewQuery(TableView.java:120)"
  "org.h2.table.TableView.initColumnsAndTables(TableView.java:172)"
  "org.h2.table.TableView.init(TableView.java:113)"
  "org.h2.table.TableView.<init>(TableView.java:70)"
  "org.h2.table.TableView.createTempView(TableView.java:559)"
  "org.h2.command.Parser.readTableFilter(Parser.java:1367)"
  "org.h2.command.Parser.parseSelectSimpleFromPart(Parser.java:2138)"
  "org.h2.command.Parser.parseSelectSimple(Parser.java:2287)"
  "org.h2.command.Parser.parseSelectSub(Parser.java:2133)"
  "org.h2.command.Parser.parseSelectUnion(Parser.java:1946)"
  "org.h2.command.Parser.parseSelect(Parser.java:1919)"
  "org.h2.command.Parser.parsePrepared(Parser.java:463)"
  "org.h2.command.Parser.parse(Parser.java:335)"
  "org.h2.command.Parser.parse(Parser.java:307)"
  "org.h2.command.Parser.prepareCommand(Parser.java:278)"
  "org.h2.engine.Session.prepareLocal(Session.java:611)"
  "org.h2.engine.Session.prepareCommand(Session.java:549)"
  "org.h2.jdbc.JdbcConnection.prepareCommand(JdbcConnection.java:1247)"
  "org.h2.jdbc.JdbcPreparedStatement.<init>(JdbcPreparedStatement.java:76)"
  "org.h2.jdbc.JdbcConnection.prepareStatement(JdbcConnection.java:1135)"
  "com.mchange.v2.c3p0.impl.NewProxyConnection.prepareStatement(NewProxyConnection.java:267)"
  "--> driver.sql_jdbc.execute$fn__76548.invokeStatic(execute.clj:238)"
  "driver.sql_jdbc.execute$fn__76548.invoke(execute.clj:236)"
  "driver.sql_jdbc.execute$prepared_statement_STAR_.invokeStatic(execute.clj:256)"
  "driver.sql_jdbc.execute$prepared_statement_STAR_.invoke(execute.clj:253)"
  "driver.sql_jdbc.execute$execute_reducible_query.invokeStatic(execute.clj:389)"
  "driver.sql_jdbc.execute$execute_reducible_query.invoke(execute.clj:376)"
  "driver.sql_jdbc.execute$execute_reducible_query.invokeStatic(execute.clj:385)"
  "driver.sql_jdbc.execute$execute_reducible_query.invoke(execute.clj:376)"
  "driver.sql_jdbc$fn__78126.invokeStatic(sql_jdbc.clj:49)"
  "driver.sql_jdbc$fn__78126.invoke(sql_jdbc.clj:47)"
  "driver.h2$fn__76938.invokeStatic(h2.clj:87)"
  "driver.h2$fn__76938.invoke(h2.clj:84)"
  "query_processor.context$executef.invokeStatic(context.clj:59)"
  "query_processor.context$executef.invoke(context.clj:48)"
  "query_processor.context.default$default_runf.invokeStatic(default.clj:69)"
  "query_processor.context.default$default_runf.invoke(default.clj:67)"
  "query_processor.context$runf.invokeStatic(context.clj:45)"
  "query_processor.context$runf.invoke(context.clj:39)"
  "query_processor.reducible$pivot.invokeStatic(reducible.clj:34)"
  "query_processor.reducible$pivot.invoke(reducible.clj:31)"
  "query_processor.middleware.mbql_to_native$mbql__GT_native$fn__47457.invoke(mbql_to_native.clj:26)"
  "query_processor.middleware.check_features$check_features$fn__46732.invoke(check_features.clj:42)"
  "query_processor.middleware.optimize_datetime_filters$optimize_datetime_filters$fn__47622.invoke(optimize_datetime_filters.clj:133)"
  "query_processor.middleware.auto_parse_filter_values$auto_parse_filter_values$fn__45533.invoke(auto_parse_filter_values.clj:44)"
  "query_processor.middleware.wrap_value_literals$wrap_value_literals$fn__49188.invoke(wrap_value_literals.clj:149)"
  "query_processor.middleware.annotate$add_column_info$fn__45295.invoke(annotate.clj:575)"
  "query_processor.middleware.permissions$check_query_permissions$fn__46607.invoke(permissions.clj:70)"
  "query_processor.middleware.pre_alias_aggregations$pre_alias_aggregations$fn__48140.invoke(pre_alias_aggregations.clj:40)"
  "query_processor.middleware.cumulative_aggregations$handle_cumulative_aggregations$fn__46805.invoke(cumulative_aggregations.clj:61)"
  "query_processor.middleware.resolve_joined_fields$resolve_joined_fields$fn__48377.invoke(resolve_joined_fields.clj:36)"
  "query_processor.middleware.resolve_joins$resolve_joins$fn__48696.invoke(resolve_joins.clj:183)"
  "query_processor.middleware.add_implicit_joins$add_implicit_joins$fn__41020.invoke(add_implicit_joins.clj:254)"
  "query_processor.middleware.large_int_id$convert_id_to_string$fn__47418.invoke(large_int_id.clj:44)"
  "query_processor.middleware.limit$limit$fn__47443.invoke(limit.clj:38)"
  "query_processor.middleware.format_rows$format_rows$fn__47398.invoke(format_rows.clj:84)"
  "query_processor.middleware.desugar$desugar$fn__46871.invoke(desugar.clj:22)"
  "query_processor.middleware.binning$update_binning_strategy$fn__45892.invoke(binning.clj:229)"
  "query_processor.middleware.resolve_fields$resolve_fields$fn__46409.invoke(resolve_fields.clj:24)"
  "query_processor.middleware.add_dimension_projections$add_remapping$fn__40564.invoke(add_dimension_projections.clj:318)"
  "query_processor.middleware.add_implicit_clauses$add_implicit_clauses$fn__40771.invoke(add_implicit_clauses.clj:141)"
  "query_processor.middleware.add_source_metadata$add_source_metadata_for_source_queries$fn__41169.invoke(add_source_metadata.clj:105)"
  "query_processor.middleware.reconcile_breakout_and_order_by_bucketing$reconcile_breakout_and_order_by_bucketing$fn__48337.invoke(reconcile_breakout_and_order_by_bucketing.clj:98)"
  "query_processor.middleware.auto_bucket_datetimes$auto_bucket_datetimes$fn__45480.invoke(auto_bucket_datetimes.clj:125)"
  "query_processor.middleware.resolve_source_table$resolve_source_tables$fn__46456.invoke(resolve_source_table.clj:46)"
  "query_processor.middleware.parameters$substitute_parameters$fn__48122.invoke(parameters.clj:114)"
  "query_processor.middleware.resolve_referenced$resolve_referenced_card_resources$fn__46508.invoke(resolve_referenced.clj:80)"
  "query_processor.middleware.expand_macros$expand_macros$fn__47127.invoke(expand_macros.clj:158)"
  "query_processor.middleware.add_timezone_info$add_timezone_info$fn__41200.invoke(add_timezone_info.clj:15)"
  "query_processor.middleware.splice_params_in_response$splice_params_in_response$fn__49058.invoke(splice_params_in_response.clj:32)"
  "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__48348$fn__48352.invoke(resolve_database_and_driver.clj:33)"
  "driver$do_with_driver.invokeStatic(driver.clj:61)"
  "driver$do_with_driver.invoke(driver.clj:57)"
  "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__48348.invoke(resolve_database_and_driver.clj:27)"
  "query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__47345.invoke(fetch_source_query.clj:267)"
  "query_processor.middleware.store$initialize_store$fn__49067$fn__49068.invoke(store.clj:11)"
  "query_processor.store$do_with_store.invokeStatic(store.clj:46)"
  "query_processor.store$do_with_store.invoke(store.clj:40)"
  "query_processor.middleware.store$initialize_store$fn__49067.invoke(store.clj:10)"
  "query_processor.middleware.cache$maybe_return_cached_results$fn__46385.invoke(cache.clj:214)"
  "query_processor.middleware.validate$validate_query$fn__49076.invoke(validate.clj:10)"
  "query_processor.middleware.normalize_query$normalize$fn__47470.invoke(normalize_query.clj:22)"
  "query_processor.middleware.add_rows_truncated$add_rows_truncated$fn__41038.invoke(add_rows_truncated.clj:36)"
  "query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__49043.invoke(results_metadata.clj:147)"
  "query_processor.middleware.constraints$add_default_userland_constraints$fn__46748.invoke(constraints.clj:42)"
  "query_processor.middleware.process_userland_query$process_userland_query$fn__48211.invoke(process_userland_query.clj:136)"
  "query_processor.middleware.catch_exceptions$catch_exceptions$fn__46691.invoke(catch_exceptions.clj:174)"
  "query_processor.reducible$async_qp$qp_STAR___39827$thunk__39828.invoke(reducible.clj:103)"
  "query_processor.reducible$async_qp$qp_STAR___39827.invoke(reducible.clj:109)"
  "query_processor.reducible$sync_qp$qp_STAR___39836$fn__39839.invoke(reducible.clj:135)"
  "query_processor.reducible$sync_qp$qp_STAR___39836.invoke(reducible.clj:134)"
  "query_processor$process_userland_query.invokeStatic(query_processor.clj:237)"
  "query_processor$process_userland_query.doInvoke(query_processor.clj:233)"
  "query_processor$fn__49234$process_query_and_save_execution_BANG___49243$fn__49246.invoke(query_processor.clj:249)"
  "query_processor$fn__49234$process_query_and_save_execution_BANG___49243.invoke(query_processor.clj:241)"
  "query_processor$fn__49278$process_query_and_save_with_max_results_constraints_BANG___49287$fn__49290.invoke(query_processor.clj:261)"
  "query_processor$fn__49278$process_query_and_save_with_max_results_constraints_BANG___49287.invoke(query_processor.clj:254)"
  "api.dataset$fn__52603$fn__52606.invoke(dataset.clj:59)"
  "query_processor.streaming$streaming_response_STAR_$fn__37131$fn__37132.invoke(streaming.clj:73)"
  "query_processor.streaming$streaming_response_STAR_$fn__37131.invoke(streaming.clj:72)"
  "async.streaming_response$do_f_STAR_.invokeStatic(streaming_response.clj:66)"
  "async.streaming_response$do_f_STAR_.invoke(streaming_response.clj:64)"
  "async.streaming_response$do_f_async$fn__24639.invoke(streaming_response.clj:85)"],
 :context :ad-hoc,
 :error
 "Column \"source.CustomExpression\" not found; SQL statement:\nCREATE FORCE VIEW PUBLIC._40 AS\nSELECT\n    \"source\".\"CustomExpression\" AS \"CustomExpression\",\n    \"source\".CREATED_AT AS CREATED_AT,\n    \"source\".\"expression\" AS \"expression\",\n    (\"source\".\"CustomExpression\" * 1234) AS \"CustomColumn\"\nFROM (\n    SELECT\n        PARSEDATETIME(YEAR(PUBLIC.ORDERS.CREATED_AT), 'yyyy') AS CREATED_AT,\n        1 AS \"expression\"\n    FROM PUBLIC.ORDERS\n    GROUP BY PARSEDATETIME(YEAR(PUBLIC.ORDERS.CREATED_AT), 'yyyy')\n    ORDER BY 1\n) \"source\" [42122-197]",
 :row_count 0,
 :running_time 0,
 :preprocessed
 {:database 4,
  :query
  {:expressions {:CustomColumn [:* [:field-literal "CustomExpression" :type/Float] 1234]},
   :source-metadata
   [{:table_id 11,
     :special_type :type/CreationTimestamp,
     :unit :year,
     :name "CREATED_AT",
     :settings nil,
     :id 86,
     :display_name "Created At",
     :fingerprint {:global {:distinct-count 10001, :nil% 0.0}, :type {:type/DateTime {:earliest "2016-05-01T00:56:13.352Z", :latest "2020-04-19T20:07:15.657Z"}}},
     :base_type :type/DateTime}
    {:name "expression", :display_name "CustomExpression", :base_type :type/Float, :special_type :type/Number}],
   :fields [[:field-literal "CREATED_AT" :type/DateTime] [:field-literal "expression" :type/Float] [:expression "CustomColumn"]],
   :limit 2000,
   :source-query
   {:source-table 11,
    :aggregation [[:aggregation-options [:* 1 1] {:display-name "CustomExpression", :name "expression"}]],
    :breakout [[:datetime-field [:field-id 86] :year]],
    :order-by [[:asc [:datetime-field [:field-id 86] :year]]]}},
  :type :query,
  :middleware {:js-int-to-string? true, :add-default-userland-constraints? true},
  :info
  {:executed-by 1,
   :context :ad-hoc,
   :nested? false,
   :query-hash [127, -65, 78, 56, -83, -85, -67, -20, -46, -57, 127, -36, 117, -61, 95, 0, 12, -30, -21, 68, 100, 22, -8, 89, -60, -29, 65, -62, -97, 122, 87, 113]},
  :constraints {:max-results 10000, :max-results-bare-rows 2000}},
 :data {:rows [], :cols []}}

2020-11-20 13:51:59,723 DEBUG middleware.log :: POST /api/dataset 202 [ASYNC: completed] 377.7 ms (8 DB calls) App DB connections: 0/10 Jetty threads: 3/50 (3 idle, 0 queued) (92 total active threads) Queries in flight: 0 (0 queued)
```

</details>

**Possible workaround** is to not create Custom Column in step 3, but simply add another metric in step 2, but that might not be an option if the Custom Column was using fields from the Group By.
![image](https://user-images.githubusercontent.com/1447303/99803135-322ef300-2b39-11eb-9fed-c0a91df5b824.png)

**Information about your Metabase Installation:**
Tested 0.34.3 thru 0.37.2

**Additional context**
Likely a duplicate of #12839 - related to #13649

**Comment by alekseyp** [comment]
I can confirm, and it's affecting what I'm trying to do a lot.

The same happens if I do multiple custom expressions and then would like to sum them up as well (originally tried as a workaround to custom column)

Let me know if I should create a new issue. But I think it's related to this one.

![Uploading Screen Shot 2020-12-01 at 8.51.21 PM.png…]()

**Comment by alekseyp** [comment]
Sorry to bother you guys, is there any ETA on that, since we really relay a lot on those, and users can't use UI to build reports for quite some time now.

**Comment by ariya** [comment]
Most likely, this is related to #13175 as well.

**Comment by flamber** [comment]
Fixed by #16259
